crates/cacache/src/content/write.rs

```diff
@@ -60,7 +112,25 @@ impl Writer {
             // Safe unwrap. cpath always has multiple segments
             .create(cpath.parent().unwrap())
             .to_internal()?;
-        let res = self.tmpfile.persist(&cpath).to_internal();
+
+        let mut maybe_mmap = self.target.into_inner().to_internal()?;
+
+        match self.expected_size {
+            None => {
+                maybe_mmap.seek(std::io::SeekFrom::Start(0)).to_internal()?;
+                let bytes = (self.written as u64).to_be_bytes();
+
+                maybe_mmap.write(&bytes).to_internal()?;
+                maybe_mmap.flush().to_internal()?;
+            }
+            Some(size) => {
+                if size != self.written {
+                    return Err(crate::errors::Error::SizeError(size, self.written));
+                }
+            }
+        };
```

Comment:
Backtrack and write the real size if we weren't given a size upfront, otherwise bail if the written size doesn't match the advertised size.